crates/bevy_tasks/src/task_pool.rs

```diff
@@ -0,0 +1,280 @@
+                            }
+
+                            if !ticker.tick() {
+                                p.park();
+                            }
+                        }
+                    })
+                    .expect("failed to spawn thread");
+
+                (handle, unparker)
+            })
+            .collect();
+
+        Self {
+            executor,
+            inner: Arc::new(TaskPoolInner {
+                threads,
+                shutdown_flag,
+            }),
+        }
+    }
+
+    /// Return the number of threads owned by the task pool
+    pub fn thread_num(&self) -> usize {
+        self.inner.threads.len()
+    }
+
+    /// Allows spawning non-`static futures on the thread pool. The function takes a callback,
+    /// passing a scope object into it. The scope object provided to the callback can be used
+    /// to spawn tasks. This function will await the completion of all tasks before returning.
+    ///
+    /// This is similar to `rayon::scope` and `crossbeam::scope`
+    pub fn scope<'scope, F, T>(&self, f: F) -> Vec<T>
+    where
+        F: FnOnce(&mut Scope<'scope, T>) + 'scope + Send,
+        T: Send + 'static,
+    {
+        let executor: &'scope multitask::Executor = unsafe { mem::transmute(&*self.executor) };
+
+        let fut = async move {
+            let mut scope = Scope {
+                executor,
+                spawned: Vec::new(),
+            };
+
+            f(&mut scope);
+
+            let mut results = Vec::with_capacity(scope.spawned.len());
+            for task in scope.spawned {
+                results.push(task.await);
+            }
+
+            results
+        };
+
+        pin_mut!(fut);
+
+        // let fut: Pin<&mut (dyn Future<Output=()> + Send)> = fut;
+        let fut: Pin<&'static mut (dyn Future<Output = Vec<T>> + Send + 'static)> =
+            unsafe { mem::transmute(fut as Pin<&mut (dyn Future<Output = Vec<T>> + Send)>) };
+
+        pollster::block_on(self.executor.spawn(fut))
+    }
+
+    /// Spawns a static future onto the thread pool. The returned Task is a future. It can also be
+    /// cancelled and "detached" allowing it to continue running without having to be polled by the
+    /// end-user.
+    pub fn spawn<T>(
+        &self,
+        future: impl Future<Output = T> + Send + 'static,
+    ) -> impl Future<Output = T> + Send
+    where
+        T: Send + 'static,
+    {
+        self.executor.spawn(future)
+    }
+}
+
+impl Default for TaskPool {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+pub struct Scope<'scope, T> {
+    executor: &'scope multitask::Executor,
+    spawned: Vec<multitask::Task<T>>,
+}
+
+impl<'scope, T: Send + 'static> Scope<'scope, T> {
+    pub fn spawn<Fut: Future<Output = T> + 'scope + Send>(&mut self, f: Fut) {
+        let fut: Pin<Box<dyn Future<Output = T> + 'scope + Send>> = Box::pin(f);
+        let fut: Pin<Box<dyn Future<Output = T> + 'static + Send>> = unsafe { mem::transmute(fut) };
```

Comment:
this is some evil rust right here :smiling_imp:

Reply:
Yeah, I'll add more comments - unsafe code deserves comments.